### PR TITLE
Conan Dependency Manager

### PR DESCRIPTION
**Pull request by pivotal-mike**
As Greenplum continues to evolve we see a growing need to be able to specify and version dependencies along with the code.  At the moment ORCA is the primary beneficiary of this but we do see several other components that have clear demarcation that would also benefit from being quasi-independent of the server code yet is still part of Greenplum.  We also would like to have a means to handle dependencies for which we are not guaranteed to be handled by the operating system.  Because of this we have opted to use Conan (http://conan.io) which is a C/C++ package manager and will be hosting packages in a dedicated Greenplum repository.  Some of the highlights of this are:

- Hosted packages with pre-build binaries
- Ability to compile packages from source when binaries are not available or if desired for testing other options
- Non-onerours requirements (requires Python 2.7 or higher)
- Agnostics of build systems - does not require retooling of existing build systems and is able to work with packages which have a variety of build mechanism
- Dependencies are versioned with the code, allowing for consistent builds
- Allows for multi-level dependencies and inclusion of all required files.

I have also create a new repository to host package definitions which is open to folks to contribute to and can be found at http://github.com/greenplum-db/conan.

### CODE REVIEW

File: depends/conanfile.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[requires]
+orca/v2.32.0@gpdb/stable
+
```
Comment (yydzero) [inline depends/conanfile.txt]: Great movement!

One question: conan will install all dependencies, no matter use --enable-orca or --disable-orca, right?  Not a big deal as it consumes only a bit more disk space, just curious.

**Comment by pivotal-mike** [comment]
That’s right however we can keep multiple files in here.  One setup could be an ORCA conanfile which would bring in just orca and another file with any other dependencies we want.  In this case the trade off is needing to run `conan import` twice but then we can separate orca from others.

~ mike

**Comment by vraghavan78** [comment]
Dumb question, does it automatically know gporca has a CMakeList.txt and understand that cmake is the way to install? Where is the xerces dependency mentioned?

**Comment by hardikar** [comment]
@vraghavan78 Take a look at https://github.com/greenplum-db/conan/blob/master/orca/orca-v2.32.0.py.

**Comment by vraghavan78** [comment]
Thanks @hardikar makes sense now.

**Comment by vraghavan78** [comment]
@pivotal-mike so we have the orca ci-pipeline to auto generate the orca-v2.XX.0.py file whenever we bump up the Orca version number.

File: depends/conanfile.txt (hunk 1 above)
Comment (vraghavan78) [inline depends/conanfile.txt]: The repro is small, and builds fast enough. So should not be much overhead.

Comment (vraghavan78) [inline depends/conanfile.txt]: @pivotal-mike can we have -j option for cmake to make the compiling faster?

**Comment by pivotal-mike** [comment]
That would probably be the easiest way to do this.  Someone will still need to bump the version as well in the GPDB Conan file as well.

**Comment by pivotal-mike** [comment]
You can add it to the orca and gpxerces build commands in their respective formulas.

File: depends/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+Conan requires python 2.7 or higher to run.  To install Conan run the command:
+
+```
+	pip install conan
+```
+
+This will install all the required python packages to use Conan on your system.  Once Conan is installed, please add the Greenplum remote Conan repository.  This is where we are hosting the packages and the pre-build binaries for GPDB and associated projects.  Anonymous access is supported for package downloading and to configure access please run the command:
+
+```
+	conan remport add <REMOTE_NAME> ttps://api.bintray.com/conan/greenplum-db/gpdb-oss
```
Comment (hsyuan) [inline depends/README.md]: s/ttps/https/

Comment (hsyuan) [inline depends/README.md]: s/remport/remote/

File: depends/conanfile.txt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+[requires]
+orca/v2.32.0@gpdb/stable
```
Comment (hsyuan) [inline depends/conanfile.txt]: currently orca v2.34 is required

**Comment by hsyuan** [comment]
Is this only for oss build? Will enterprise build benefit from this?

**Comment by hsyuan** [comment]
@vraghavan78 since gp-xerces is not required anymore, can we prioritize it to remove dependency on it sooner than later? installing from apache official binary should suffice.

**Comment by vraghavan78** [comment]
@hsyuan this is a dev issue to prioritize accordingly

**Comment by vraghavan78** [comment]
@d @oarap please take a look and get this to a closure

File: depends/conanfile.txt (hunk 1 above)
Comment (pivotal-mike) [inline depends/conanfile.txt]: @vraghavan78 Yes - though we would need to define that in the ORCA recipe.  Maybe a walkthrough of that with the folks next week might make sense

File: depends/conanfile.txt (hunk 3 above)
Comment (pivotal-mike) [inline depends/conanfile.txt]: At time of writing 2.32 was the available version.  we need to hand this off to the folks working on ORCA to keep up to date

**Comment by pivotal-mike** [comment]
@hsyuan The idea is that this will be used for both the OpenSource and the builds that Pivotal publishes

**Comment by hsyuan** [comment]
Any updates on this PR?

**Comment by Eulerizeit** [comment]
@pivotal-mike is there something I can do to help push this through?

**Comment by pivotal-mike** [comment]
@Eulerizeit fyi

**Comment by pivotal-mike** [comment]
@oarap @vraghavan78 FYI - merged in Conan, we can start looking at how to vendor ORCA for the open source builds now

**Pull request merged**